### ISSUE TITLE
Reuse existing poetry env for server if available

### ISSUE THREAD

**Issue by timokau** [timokau/prefq#48]
Currently the `run_server.sh` script installs a poetry environment without any extras, while `run_imitation_example.sh` installs an environment with `-E examples`. As a result, the environment has to be re-installed all the time.

The server can run just fine with the `-E examples` extras, it doesn't hurt to have them. At the same time we do not want to require them. It would be best if we could check whether the `-E examples` env is already active. In that case the server could use it, if not it could install a new environment (without any extras).

**Comment by patrickab** [comment]
This turns out a bit more difficult than one may expect. I tried my luck with `poetry show -t` and `poetry env info`, unfortunately theres no easy built-in tool to achieve this. I've also tried to find a way to merge all examples into one package, so that they get listed in the same branch hierarchy whenever i call `poetry show -t`, this way it would be possible to just grep for the `extras package`.

On stackoverflow [I found](https://stackoverflow.com/questions/68203376/how-to-list-the-name-of-all-poetry-extras) someone with a similar problem, where the only functional suggestion is to use a `sed` statement, to extract all packages contained in extras.

**Comment by timokau** [comment]
Can we maybe use `poetry install --dry-run` to check if the extras are already available?

**Comment by patrickab** [comment]
after a slight misunderstanding about `poetry install --dry-run`, I've found a much easier solution

inside the shellscripts we can simply run poetry run `python3 -m prefq.<python_module>` inside an if condition. By negating the operation we can `poetry install` only if necessary, otherwise we simply run the script without entering the condition.

```bash
# run "poetry install" only if dependencies are missing
if ! poetry run python3 -m prefq.server "$@"; then
    echo "...installing missing dependencies"
    poetry install
    poetry run python3 -m prefq.server "$@"
fi
```

**Comment by timokau** [comment]
I don't think this works since if I recall correctly `poetry run` does not fail when the environment is not yet installed. Have you tested this?

**Comment by patrickab** [comment]
poetry run raises an `ModuleNotFoundError`whenever any module is not contained in the currently activated virtual environment.

I had tested this on my own laptop before, however, for additional verification I've just tested this on another machine, executed from a freshly cloned repository. This solution works fine.

**Comment by patrickab** [comment]
poetry run raises a `ModuleNotFoundErrorwhenever` if any module is not contained in the currently activated virtual environment.

I had tested this on my own laptop before, however, for additional verification I've just tested this on another machine, executed from a freshly cloned repository. This solution works flawlessly.

**Comment by timokau** [comment]
Yes, but as far as I know this is not raised by poetry but from the python program itself when it cannot access a module. When e.g. a different version of a module is available, it will not complain. You could, e.g., have `package_x` at version 2 installed in your system python environment while the poetry package requires version 1. I think `poetry run` will not complain in this case and just use version 2 from the system.
